apps/wallet/src/background/Keyring.ts

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { Ed25519Keypair } from '@mysten/sui.js';
+import { EventEmitter } from 'events';
+import Browser from 'webextension-polyfill';
+
+import { encrypt, decrypt } from '_shared/cryptography/keystore';
+import { generateMnemonic } from '_shared/utils/bip39';
+
+import type { Keypair } from '@mysten/sui.js';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type ListenerFn = (...args: any[]) => void;
+
+export enum KeyringEvent {
+    lockedStatusUpdate = 'lockedStatusUpdate',
+}
+
+const STORAGE_KEY = 'vault';
+
+class Keyring {
+    #events = new EventEmitter();
+    #locked = true;
+    #keypair: Keypair | null = null;
+    #mnemonic: string | null = null;
+
+    // Creates a new mnemonic and saves it to storage encrypted
+    public async createMnemonic(password: string) {
+        if (await this.isWalletInitialized()) {
+            throw new Error(
+                'Mnemonic already exists, creating a new one will override it. Clear the existing one first.'
+            );
+        }
+        const encryptedMnemonic = await encrypt(
+            password,
+            Buffer.from(generateMnemonic(), 'utf8')
+        );
+        await this.storeEncryptedMnemonic(encryptedMnemonic);
+    }
+
+    public lock() {
+        this.#keypair = null;
+        this.#mnemonic = null;
+        this.#locked = true;
+        this.notifyLockedStatusUpdate(this.#locked);
+    }
+
+    public async unlock(password: string) {
+        this.#mnemonic = await this.decryptMnemonic(password);
+        this.#keypair = Ed25519Keypair.deriveKeypair(this.#mnemonic);
+        this.#locked = false;
+        this.notifyLockedStatusUpdate(this.#locked);
+    }
+
+    public async clearMnemonic() {
+        await this.storeEncryptedMnemonic(null);
+        this.lock();
+    }
+
+    public async isWalletInitialized() {
+        return !!(await this.loadMnemonic());
+    }
+
+    public get isLocked() {
+        return this.#locked;
+    }
+
+    public get keypair() {
+        return this.#keypair;
+    }
+
+    public get mnemonic() {
```

Comment:
It seems like this should be a function that needs to intentionally be called, rather than something that can be accessed like this.

Reply:
Done